src/pages/home/report/withReportOrNavigateHome.js

```diff
@@ -0,0 +1,59 @@
+import PropTypes from 'prop-types';
+import React, {Component} from 'react';
+import {withOnyx} from 'react-native-onyx';
+import _ from 'underscore';
+import getComponentDisplayName from '../../../libs/getComponentDisplayName';
+import Navigation from '../../../libs/Navigation/Navigation';
+import ONYXKEYS from '../../../ONYXKEYS';
+import reportPropTypes from '../../reportPropTypes';
+
+export default function (WrappedComponent) {
+    const propTypes = {
+        /** The HOC takes an optional ref as a prop and passes it as a ref to the wrapped component.
+          * That way, if a ref is passed to a component wrapped in the HOC, the ref is a reference to the wrapped component, not the HOC. */
+        forwardedRef: PropTypes.func,
+
+        /** The report currently being looked at */
+        report: reportPropTypes,
+    };
+
+    const defaultProps = {
+        forwardedRef: () => {},
+        report: {},
+    };
+
+    class WithReportOrNavigateHome extends Component {
```

Comment:
Ouch, did we really need an HOC for this? Very confused about what purpose this even serves. @Beamanator would you mind summarizing why we need this?

Reply:
Is your question "Why a HOC for this functionality" or "What is the point of the code in this HOC"? Will happily respond to either one tomorrow (end of my day & brain is foggy from timezone shifting)

Reply:
Yes to both.

Reply:
"What is the point of the code in this HOC" should be pretty straightforward: If the user is on a page that needs `report` data (from Onyx), we check if `this.props.report` is empty or not. If it's empty, we dismiss the open modal (this is what we were calling "navigate home" though that's not really accurate since we don't have a "home" screen...). If it's not empty, we show that page.

"Why a HOC for this functionality": We wanted to use this functionality across multiple pages, so thought it was clean to put it in an HOC. This lets us not have to subscribe to `${ONYXKEYS.COLLECTION.REPORT}${route.params.reportID}`, on multiple pages, and instead we just wrap the page with this new HOC. We could do the same thing with a new `action` or `lib` that we call whenever these pages are mounted, but in my brain an HOC seemed more useful.

Reply:
Cool, thanks and sorry for my initial reaction. I think we got the main solution for this wrong which is the important thing. I tend to only pull out the HOC when we the code reuse is expected to be very high for a solution. But if you thought we might re-use this then the HOC is certainly one way to do that.

Reply:
> I think we got the main solution for this wrong which is the important thing

I think you're referring to the fact that we aren't actually sure if we want to navigate at this point, right? Like we probably want to see if the user has access to that report or not before navigating away? (this is what our other discussion is about in a different thread, I believe)

> I tend to only pull out the HOC when we the code reuse is expected to be very high for a solution. But if you thought we might re-use this then the HOC is certainly one way to do that.

Yeah exactly that was my thought process too, **BUT** maybe you would have waited till we have a few more use cases for this kind of code before making it a HOC - I'll happily defer to you on this, though I guess the "next steps" probably depend on what the "correct solution" is (again referring to our other discussion - showing a "you don't have accesss" page or navigating somewhere)

Reply:
>we aren't actually sure if we want to navigate at this point, right

Yes. I think there are different possible states of report access and different app states depending on if those reports were previously accessible or not.

Using the existence of the report in storage doesn't tell us whether you have "access" or not especially when you are offline...

**Offline and never attempted to fetch a report:** No way to _really_ tell if the user has access or not but that's fine the same error message works in this case.
**Offline and previous attempt returned 404:** Tell them they can't access it i.e. "Hmm... it's not here"
**Online and attempt returns 404:** Tell them they can't access it i.e. "Hmm... it's not here"

I don't think there is any situation where we'd want to redirect without any explanation (but this deserves a wider discussion to get more input than just me 😅).